商品出品機能

kiinoharu opened the pull request:
# What
①商品出品ページへの遷移制御
・ログイン状態でのみ商品出品ページにアクセスでき、ログアウト状態ではログインページにリダイレクトされる。
②商品情報の入力と保存
・必須項目（商品名、商品画像、商品の説明、カテゴリー、商品の状態、配送料の負担、発送元の地域、発送までの日数、価格）の入力を必須化。
・商品画像の1枚添付が必須。
カテゴリーや状態などの選択項目はデフォルトで「---」を初期値として設定。
・価格は¥300~¥9,999,999の間で半角数値のみ受け付けるようバリデーションを追加。
③販売手数料・利益計算
・価格入力時に、販売手数料（10%）と利益が自動で計算され、小数点以下を切り捨てて表示。
④エラーハンドリング
・必須項目が未入力の場合、適切なエラーメッセージを表示し、入力済み項目（商品画像・手数料・利益以外）は保持される。
・重複したエラーメッセージは表示されない。
⑤商品情報保存後の遷移
・正常に出品が完了すると、トップページにリダイレクト。

# Why
　この機能は、アプリケーションで商品の出品を可能にするための主要な部分であり、ユーザーが簡単かつ安全に商品を登録できるようにするために重要です。また、バリデーションとエラーハンドリングにより、ユーザーの入力ミスを防ぐとともに、利便性の高いユーザーエクスペリエンスを提供します。

# プルリクエストへ記載するgyazo
・ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/983b6e562589246191979e640706b7ba
・価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/4c82794f953809235a76c90b8a0d5d9d
・必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/d510cf12074698c40418ccf29963ac90
https://gyazo.com/7bbec548a63f992c2f75c18b00d54440
・入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/611b6f12795b5845dbab0585c12ae87f
・ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/160c3e00767146dc3a6f9bc4646ec245
・テスト結果の画像
https://gyazo.com/20513874c531ab27d0e214ba254660fc

kiinoharu committed:
itemモデル作成コマンド実行

kiinoharu committed:
テーブル作成、db:migrate実行

kiinoharu committed:
item-user間のアソシエーション設定

kiinoharu committed:
newアクション及びルーティング設定

kiinoharu committed:
トップページの「出品する」ボタンの遷移先設定

kiinoharu committed:
トップページの「出品する」ボタンの遷移先がログイン状態によって変化するよう設定

kiinoharu committed:
Active Storage導入

kiinoharu committed:
暫定バリデーション設定

kiinoharu committed:
ActiveHash導入

kiinoharu committed:
Categoryのデータを記述

kiinoharu committed:
ActiveHash、categoryに関わるアソシエーション設定

kiinoharu committed:
ActiveHash、categoryに関するバリデーション記述

kiinoharu committed:
category_idを含めたデータの保存処理

kiinoharu committed:
商品出品画面の編集（from_with,item-image,item-name,item_description,category_id）

from_with,item-image,item-name,item_description,category_idについて編集。
その他ActiveHashについては、以後記述。

kiinoharu committed:
createアクション修正

kiinoharu committed:
tem_conditionモデル作成

kiinoharu committed:
prefectureモデル作成

kiinoharu committed:
shipping_dayモデル作成

kiinoharu committed:
shipping_fee_burdenモデル作成

kiinoharu committed:
iteｍモデル編集

ActiveHash関連のバリデーション編集＋アソシエーション設定

kiinoharu committed:
itemsコントローラー編集

permitにActiveHash関連の記述追記

kiinoharu committed:
商品出品ページ編集

ActiveHashを反映させる用記述

kiinoharu committed:
createアクションに対するルーティング設定

kiinoharu committed:
itemモデル編集

priceカラムのバリデーション追記

kiinoharu committed:
JavaScriptを用いて販売手数料、販売利益の表示が変わることよう設定

kiinoharu committed:
商品出品画面、priceのname属性設定

kiinoharu committed:
itemsコントローラー及び商品出品画面記述修正

kiinoharu committed:
itemsコントローラー修正

permitの記述順整理

kiinoharu committed:
商品出品ページ修正

from_withのurl削除。
Itemモデルに基づいて正しいURL(/items)にPOSTリクエストを送信するため。

kiinoharu committed:
商品出品ページ編集

エラーハンドリングの部分テンプレート読み込み

kiinoharu committed:
iteｍsコントローラー編集

merge追記

kiinoharu committed:
itemモデル単体テスト用サンプル画像追加

kiinoharu committed:
iteｍモデル単体テスト用FactoryBot実装

kiinoharu committed:
iteｍモデル単体テストコード実装

kiinoharu committed:
itemモデル修正

priceカラムのバリデーション修正

motomuun commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,25 @@
 class ItemsController < ApplicationController
```
ログアウト状態のユーザーが商品出品できないための設定がコントローラーに記述されていません。
以下の実装要件を満たすよう、必要な設定をbefore_actionで指定しましょう！
もし、applicationコントローラーに記述している場合は、itemsコントローラーに記述をお願いします！

```
・ ログイン状態のユーザーのみ、商品出品ページへ遷移できること
・ログアウト状態のユーザーは、商品出品ページへ遷移しようとすると、ログインページへ遷移すること
```

motomuun commented on app/javascript/packs/items.js:
```diff
@@ -0,0 +1,24 @@
+      profit.textContent = '';
+    }
+  });
+});
```
renderメソッドに対応したイベントを追加しましょう。
以下のカリキュラムをご参考になさってください。
```
最終課題補足カリキュラム | Lesson2 実装に悩んだときのヒント
STEP2 手順6  renderメソッドに対応したイベントを追加する
```

motomuun commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,89 @@
+        @item.price = '１０００'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is not a number')
+      end
```
userが紐づいていないと保存できないことを確認するテストコードを実装しましょう！

motomuun reviewed (commented):
レビュー依頼ありがとうございます！
修正点をコメントさせて頂きましたので、ご確認をお願い致します。
ご不明点や疑問点がありましたら、チャット・コネクトメンターにご相談ください。

【次回プルリクエスト提出する際の注意点】
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。

**※お手数ではありますがテストコードを追加した後は、改めてテストコード実行後のキャプチャ画像を撮影いただき、こちらのプルリクエストへ添付していただくようお願いいたします。**

kiinoharu committed:
itemsコントローラー修正

before_action :authenticate_user!を用いて、ログアウト状態のユーザーが商品出品ページへアクセスしようとすると、ログインページにリダイレクトされるよう設定。

kiinoharu committed:
items.js修正

記述内容変更＋ renderメソッドに対応したイベントを追加

kiinoharu committed:
itemモデル単体テストコード追加

'userが紐づいていなければ出品できない'を追加

kiinoharu commented:
レビューいただき、ありがとうございます。
ご指摘いただいた内容について、修正致しました。
再度、ご確認の程よろしくお願い致します。

・テストコード修正後のテスト成功画像
https://gyazo.com/3f016cf0467e2fafff3855ac294c1067

motomuun commented:
レビューのご依頼ありがとうございます。

実装内容を確認させていただきました。
**LGTMとさせていただきます。実装お疲れ様でした！**

このブランチをマージし、masterブランチ（もしくはmainブランチ）でプルしましょう。
内容が反映されていることを確認し、次の実装・学習に入りましょう！
引き続き応援しております！
